# **Enumeration of q-matroids (Dimension 3,4 and 5)**
***
*** 

**In this notebook we want to present an enumeration algorithm of q-matroids in dimensions 3,4 and 5.**

**The dimension 3 case is already well known, but we present anyways, since it is the easiest and quickest example of the algorithm.**

**For the dimension 4 and 5 case we only enumerate the q-matroids of rank 2, since all others are either dual cases of rank 2 or dual cases of rank 1 (which are given by choosing the loop-space).**

**Moreover we only consider vector-spaces over the finite field with characteristic 2.**

## Imports
***

In [1]:
using Oscar
import Oscar: ideal
using Combinatorics
using Graphs
using Graphs.Experimental 
using GraphPlot
using Compose
using BSON
using Random
using InvertedIndices
using DataStructures
include("./helper_functions.jl")
include("./q_properties.jl")
include("./q_matroids.jl")
include("./enumeration.jl")

Dim5_q_matroid_DFS (generic function with 1 method)

## Dimension 3
***

**We start by taking a q-matroid in dimension 2 and embed this one in dimension 3 by setting the third coordinate to 0.**

**So since we looking for 3-dim. q-matroids of rank 2 we have 5 possible options for this initial 2-dim. q-matroid to embed.**

**Moreover since all 2-dim. q-matroids are representable we only state the matrix representing the q-matroid:**

1. $\Big(\begin{matrix}1& 0\end{matrix}\Big)_{\mathbb{F}_2}$ = rank 1 q-matroid with 1-dim. loopspace
2. $\Big(\begin{matrix}0& 1\end{matrix}\Big)_{\mathbb{F}_2}$ = rank 1 q-matroid with 1-dim. loopspace
3. $\Big(\begin{matrix}1& 1\end{matrix}\Big)_{\mathbb{F}_2}$ = rank 1 q-matroid with 1-dim. loopspace
4. $\Big(\begin{matrix}1& x\end{matrix}\Big)_{\mathbb{F}_2}$ = rank 1 Uniform q-matroid
5. $\Big(\begin{matrix}1& 0\\0& 1\end{matrix}\Big)_{\mathbb{F}_{2^2}}$ = rank 2 Uniform q-matroid

**Here is one example:**

In [39]:
Ext_F,x = FiniteField(2,2,"x")
A = matrix(Ext_F,[1 x])
QM = q_matroid_from_matrix(Ext_F,A)

Q-Matroid of rank 1 in 2-dim. vector-space over the Finite field of characteristic 2

In [40]:
indeps = Q_Matroid_Indepentspaces(QM)
deps = Q_Matroid_Depentspaces(QM)
indeps,deps

(fpMatrix[[0 0], [1 0], [0 1], [1 1]], fpMatrix[[1 0; 0 1]])

In [41]:
triples = Dim3_q_matroid_DFS(QM)

4-element Vector{Any}:
 Vector{fpMatrix}[[[0 0 0], [1 0 0], [0 1 0], [1 1 0], [0 0 1], [1 0 1], [0 1 1], [1 1 1], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]], [[1 0 0; 0 1 0], [1 0 0; 0 1 0; 0 0 1]], [[1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]]
 Vector{fpMatrix}[[[0 0 0], [1 0 0], [0 1 0], [1 1 0], [0 0 1], [1 0 1], [0 1 1], [1 1 1], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 0; 0 1 1]], [[1 0 0; 0 1 0], [1 0 0; 0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 1]], [[1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 0; 0 1 1]]]
 Vector{fpMatrix}[[[0 0 0], [1 0 0], [0 1 0], [1 1 0], [0 0 1], [1 0 1], [0 1 1], [1 1 1], [1 0 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 1], [1 0 0; 0 1 1]], [[1 0 0; 0 1 0], [1 0 0; 0 1 0; 0 0 1], [0 1 0; 0 0 1], [1 0 1; 0 1 0]], [[1 0 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]]
 Vector{fpMatrix}[[[0 0 0], [1 0 0], [0 1 0], [1 1 0], [0 0 

In [42]:
true_list_alt = []
false_list_alt = []
for triple in triples
    answer = Are_q_matroid_dependentspaces(QM.field,triple[2])
    if answer == true
        push!(true_list_alt,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[1],triple[2]])
    else
        push!(false_list_alt,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[2]])
    end
end
true_list_alt

1-element Vector{Any}:
 Any[6, 16, fpMatrix[[1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]], fpMatrix[[0 0 0], [1 0 0], [0 1 0], [1 1 0], [0 0 1], [1 0 1], [0 1 1], [1 1 1], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]], fpMatrix[[1 0 0; 0 1 0], [1 0 0; 0 1 0; 0 0 1]]]

**At this point we provide a list of all possible inputs and their respective outputs:**

**(After checking if the output is a q-matroid)**

| Input q-mat | Num Output | Excat splitting in terms of num. of Baseselm'S | Remark |
|:-:|:-:|:-:|:-:|
|$\Big(\begin{matrix}1& 0\end{matrix}\Big)_{\mathbb{F}_2}$|1|1 $\times$ four bases-elm's|loop-space: $\Big\langle\begin{matrix}0 & 1 & 0\end{matrix}\Big\rangle$.|
|$\Big(\begin{matrix}0& 1\end{matrix}\Big)_{\mathbb{F}_2}$|1|1 $\times$ four bases-elm's|loop-space: $\Big\langle\begin{matrix}1 & 0 & 0\end{matrix}\Big\rangle$.|
|$\Big(\begin{matrix}1& 1\end{matrix}\Big)_{\mathbb{F}_2}$|1|1 $\times$ four bases-elm's|loop-space: $\Big\langle\begin{matrix}1 & 1 & 0\end{matrix}\Big\rangle$.|
|$\Big(\begin{matrix}1& x\end{matrix}\Big)_{\mathbb{F}_{2^2}}$|1|1 $\times$ six bases-elm's|leftover two-space:  $\Big\langle\begin{matrix}1 & 0 & 0\\0 & 1 & 0\end{matrix}\Big\rangle$.|
|$\Big(\begin{matrix}1& 0\\0& 1\end{matrix}\Big)_{\mathbb{F}_2}$|7|6 $\times$ six bases-elm's,<br>1 $\times$ seven bases-elm's|every six bases-elm's q-mat has one leftover two space not being $\Big\langle\begin{matrix}1 & 0 & 0\\0 & 1 & 0\end{matrix}\Big\rangle$,<br>the seven bases-elm's q-mat is the uniform of rank 2.|

**Now we collect all these different q-matroids in one list and run an algorithm to distinguish the different isomorphism-classes:**

In [47]:
Ext_F,x = FiniteField(2,2,"x")
rep_matrices = [matrix(Ext_F,[1 0]),matrix(Ext_F,[0 1]),matrix(Ext_F,[1 1]),matrix(Ext_F,[1 x]),matrix(Ext_F,[1 0;0 1])];

In [51]:
overall_li = AbstractVector{AbstractVector{fpMatrix}}([])
for m in rep_matrices
    true_list = []
    QM = q_matroid_from_matrix(Ext_F,m)
    triples = Dim3_q_matroid_DFS(QM)
    for triple in triples
        answer = Are_q_matroid_dependentspaces(QM.field,triple[2])
        if answer == true
            push!(true_list,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[1],triple[2]])
        end
    end
    for tuple in true_list
        push!(overall_li,tuple[3])
    end
end     
overall_li    

11-element Vector{AbstractVector{fpMatrix}}:
 [[1 0 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]
 [[0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1]]
 [[1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 0; 0 1 1]]
 [[1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]
 [[1 0 0; 0 1 0], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]
 [[1 0 0; 0 1 0], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1]]
 [[1 0 0; 0 1 0], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 0; 0 1 1]]
 [[1 0 0; 0 1 0], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]
 [[1 0 0; 0 1 0], [1 0 0; 0 0 1], [0 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]
 [[1 0 0; 0 1 0], [1 0 0; 0 0 1], [1 1 0; 0 0 1], [1 0 1; 0 1 0], [1 0 1; 0 1 1], [1 0 0; 0 1 1]]
 [[1 0 0; 0 1 0], [0 1 0; 0 0 1], [1 1 0; 0 0 1], [1 0 

In [52]:
Isom_classes_from_basesV2(GF(2),overall_li)

3-element Vector{Any}:
 (Q-Matroid of rank 2 in 3-dim. vector-space over the Finite field of characteristic 2, 4, 2)
 (Q-Matroid of rank 2 in 3-dim. vector-space over the Finite field of characteristic 2, 6, 6)
 (Q-Matroid of rank 2 in 3-dim. vector-space over the Finite field of characteristic 2, 7, 0)

## Dimension 4
***

**We start by taking a q-matroid in dimension 3 and embed this one in dimension 4 by setting the fourth coordinate to 0.**

**So since we looking for 4-dim. q-matroids of rank 2 we have 6 possible options for this initial 3-dim. q-matroid to embed.**

**Unlike above, where we really considered all possible 2-dim q-matroids, here we only looking at the possible isomorphism-classes and choose a representative.**

**Moreover since all 3-dim. q-matroids are representable we only state the matrix representing a representative q-matroid of an isomorphism-class:**

1. $\Big(\begin{matrix}1& 0& 0\end{matrix}\Big)_{\mathbb{F}_2}$ = rank 1 q-matroid with 2-dim. loopspace
2. $\Big(\begin{matrix}1& 0 & x\end{matrix}\Big)_{\mathbb{F}_{2^2}}$ = rank 1 q-matroid with 1-dim. loopspace
3. $\Big(\begin{matrix}1& x & x^2\end{matrix}\Big)_{\mathbb{F}_{2^3}}$ = rank 1 Uniform q-matroid
4. $\Big(\begin{matrix}1& 0 & 0\\0& 1 & 0\end{matrix}\Big)_{\mathbb{F}_2}$ = rank 2 q-matroid with 1-dim. loopspace (four-elm. bases)
5. $\Big(\begin{matrix}1& 0 & 0\\0& 1 & x\end{matrix}\Big)_{\mathbb{F}_{2^2}}$ = rank 2 q-matroid with six-elm. bases
6. $\Big(\begin{matrix}1& 0 & x\\0& 1 & x^2\end{matrix}\Big)_{\mathbb{F}_{2^3}}$ = rank 2 Uniform q-matroid

### Isomorphism-classes from matrices

**Before we do an example of the algorithm, we want to get an idea what kind of q-matroids we could expect.**

**Therefore we generate (2 $\times$ 4)-matrices with entries in $\mathbb{F}_{2^4}$, since we that every single of those matrices will give us a 4-dim. q-matroid.**

**(Of course only representable ones)**

**After this we run an algorithm on the set of those matrices to identify all the different isomorphism-classes.**

**The results are given in the following list of the form:**

***( Rep.-Gen.-Mat. |...| # bases-elm's | # elm's in isom-class )***

```
9-element Vector{Any}:
 ([1 0 0 0; 0 1 1 x], ..., 24, 1007)
 ([1 0 0 0; 0 1 1 0], ..., 16, 15)
 ([1 0 0 x; 0 1 1 x^2], ..., 28, 2015)
 ([1 0 0 0; 0 1 x x^2], ..., 28, 2015)
 ([1 0 0 x; 0 1 x x^2], ..., 32, 6143)
 ([1 0 0 x; 0 1 x x^3], ..., 34, 21215)
 ([1 0 0 x^2; 0 1 x x^3], ..., 33, 31679)
 ([1 0 0 x^2+x; 0 1 x^2+x x^2+x+1], ..., 30, 95)
 ([1 0 x x^2; 0 1 x^2 x^2+x], ..., 35, 1343)
```

### Enumeration-Algorithm

**Here is one example:**

In [42]:
Ext_F,x = FiniteField(2,3,"x")
A = matrix(Ext_F,[1 0 x;0 1 x^2])
QM = q_matroid_from_matrix(Ext_F,A)

Q-Matroid of rank 2 in 3-dim. vector-space over the Finite field of characteristic 2

In [43]:
indeps = Q_Matroid_Indepentspaces(QM)
deps = Q_Matroid_Depentspaces(QM)
indeps,deps

(fpMatrix[[0 0 0], [1 1 0], [0 1 0], [1 0 0], [1 0 1], [0 0 1], [1 1 1], [0 1 1], [1 0 0; 0 1 0], [1 0 0; 0 0 1], [1 0 0; 0 1 1], [0 1 0; 0 0 1], [1 0 1; 0 1 0], [1 1 0; 0 0 1], [1 0 1; 0 1 1]], fpMatrix[[1 0 0; 0 1 0; 0 0 1]])

In [44]:
triples = Dim4_q_matroid_DFS(QM)

705-element Vector{Any}:
 Vector{fpMatrix}[[[0 0 0 0], [1 1 0 0], [0 1 0 0], [1 0 0 0], [1 0 1 0], [0 0 1 0], [1 1 1 0], [0 1 1 0], [0 0 0 1], [1 0 0 1]  …  [0 1 0 1; 0 0 1 1], [1 0 1 1; 0 1 0 1], [1 0 1 0; 0 1 1 1], [1 0 1 1; 0 1 1 0], [1 0 0 0; 0 0 1 1], [0 1 0 0; 0 0 1 1], [1 1 0 0; 0 0 1 1], [1 0 1 1; 0 1 0 0], [1 0 1 1; 0 1 1 1], [1 0 0 0; 0 1 1 1]], [[1 0 0 0; 0 1 0 0; 0 0 1 0], [1 0 0 0; 0 1 0 0; 0 0 1 0; 0 0 0 1], [1 0 0 0; 0 1 0 0; 0 0 0 1], [1 0 0 0; 0 1 0 0; 0 0 1 1], [1 0 0 0; 0 0 1 0; 0 0 0 1], [1 0 0 0; 0 1 0 1; 0 0 1 0], [1 0 0 0; 0 1 1 0; 0 0 0 1], [1 0 0 0; 0 1 0 1; 0 0 1 1], [0 1 0 0; 0 0 1 0; 0 0 0 1], [1 0 0 1; 0 1 0 0; 0 0 1 0]  …  [1 0 0 1; 0 1 0 0; 0 0 1 1], [1 1 0 0; 0 0 1 0; 0 0 0 1], [1 0 0 1; 0 1 0 1; 0 0 1 0], [1 0 1 0; 0 1 1 0; 0 0 0 1], [1 0 0 1; 0 1 0 1; 0 0 1 1], [0 1 1 0; 0 0 0 1], [1 0 1 0; 0 0 0 1], [1 1 0 0; 0 0 0 1], [1 1 0 1; 0 0 1 0], [1 1 0 1; 0 0 1 1]], [[1 0 0 0; 0 1 0 0], [1 0 0 0; 0 0 1 0], [1 0 0 0; 0 1 1 0], [0 1 0 0; 0 0 1 0], [1 0 1 0; 0 

In [45]:
true_list_alt = []
false_list_alt = []
for triple in triples
    answer = Are_q_matroid_dependentspaces(QM.field,triple[2])
    if answer == true
        push!(true_list_alt,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[1],triple[2]])
    else
        push!(false_list_alt,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[2]])
    end
end
true_list_alt

477-element Vector{Any}:
 Any[35, 67, fpMatrix[[1 0 0 0; 0 1 0 0], [1 0 0 0; 0 0 1 0], [1 0 0 0; 0 1 1 0], [0 1 0 0; 0 0 1 0], [1 0 1 0; 0 1 0 0], [1 1 0 0; 0 0 1 0], [1 0 1 0; 0 1 1 0], [1 0 0 0; 0 0 0 1], [0 1 0 0; 0 0 0 1], [1 1 0 0; 0 0 0 1]  …  [1 1 0 1; 0 0 1 0], [1 0 1 0; 0 1 1 1], [1 0 1 1; 0 1 1 0], [1 1 0 1; 0 0 1 1], [1 0 0 0; 0 0 1 1], [0 1 0 0; 0 0 1 1], [1 1 0 0; 0 0 1 1], [1 0 1 1; 0 1 0 0], [1 0 1 1; 0 1 1 1], [1 0 0 0; 0 1 1 1]], fpMatrix[[0 0 0 0], [1 1 0 0], [0 1 0 0], [1 0 0 0], [1 0 1 0], [0 0 1 0], [1 1 1 0], [0 1 1 0], [0 0 0 1], [1 0 0 1]  …  [1 1 0 1; 0 0 1 0], [1 0 1 0; 0 1 1 1], [1 0 1 1; 0 1 1 0], [1 1 0 1; 0 0 1 1], [1 0 0 0; 0 0 1 1], [0 1 0 0; 0 0 1 1], [1 1 0 0; 0 0 1 1], [1 0 1 1; 0 1 0 0], [1 0 1 1; 0 1 1 1], [1 0 0 0; 0 1 1 1]], fpMatrix[[1 0 0 0; 0 1 0 0; 0 0 1 0], [1 0 0 0; 0 1 0 0; 0 0 1 0; 0 0 0 1], [1 0 0 0; 0 1 0 0; 0 0 0 1], [1 0 0 0; 0 1 0 0; 0 0 1 1], [1 0 0 0; 0 0 1 0; 0 0 0 1], [1 0 0 0; 0 1 0 1; 0 0 1 0], [1 0 0 0; 0 1 1 0; 0 0 0 1], [1 0 

In [46]:
cards = counter([elm[1] for elm in true_list_alt])

Accumulator{Int64, Int64} with 5 entries:
  34 => 28
  32 => 224
  35 => 1
  33 => 168
  31 => 56

**We provide at this point again a list of all possible inputs and their respective outputs:**

**(After checking if the output is a q-matroid)**

| Input q-mat | Num Output | Excat splitting in terms of num. of Baseselm'S | Remark |
|:-:|:-:|:-:|:-:|
|$\Big(\begin{matrix}1& 0& 0\end{matrix}\Big)_{\mathbb{F}_2}$|1|1 $\times$ 16 bases-elm's|loop-space: $\Big\langle\begin{matrix}0 & 1 & 0 & 0\\0 & 0 & 1 & 0\end{matrix}\Big\rangle$.|
|$\Big(\begin{matrix}1& 0 & x\end{matrix}\Big)_{\mathbb{F}_{2^2}}$|1|1 $\times$ 24 bases-elm's|loop-space: $\Big\langle\begin{matrix}0 & 1 & 0 & 0\end{matrix}\Big\rangle$.|
|$\Big(\begin{matrix}1& x & x^2\end{matrix}\Big)_{\mathbb{F}_{2^3}}$|1|1 $\times$ 28 bases-elm's|loop-space: none, **paving**.|
|$\Big(\begin{matrix}1& 0 & 0\\0& 1 & 0\end{matrix}\Big)_{\mathbb{F}_2}$|7|6 $\times$ 24 bases-elm's,<br>1 $\times$ 28 bases-elm's |every 24 bases-elm's q-mat has a one-dim. loop-space,<br> the 28 bases-elm's q-mat has also a one-dim. loop-space (**not paving**).|
|$\Big(\begin{matrix}1& 0 & 0\\0& 1 & x\end{matrix}\Big)_{\mathbb{F}_{2^2}}$|107|2 $\times$ 28 bases-elm's,<br> 8 $\times$ 30 bases-elm's,<br> 32 $\times$ 31 bases-elm's,<br>48 $\times$ 32 bases-elm's,<br>16 $\times$ 33 bases-elm's,<br>1 $\times$ 34 bases-elm's|**...**|
|$\Big(\begin{matrix}1& 0 & x\\0& 1 & x^2\end{matrix}\Big)_{\mathbb{F}_{2^3}}$|477|56 $\times$ 31 bases-elm's,<br>224 $\times$ 32 bases-elm's,<br>168 $\times$ 33 bases-elm's,<br>28 $\times$ 34 bases-elm's,<br>1 $\times$ 35 bases-elm's|**...**|

**Now we collect all these different q-matroids in one list and run an algorithm to distinguish the different isomorphism-classes:**

In [48]:
Ext_F,x = FiniteField(2,3,"x")
rep_matrices = [matrix(Ext_F,[1 0 0]),matrix(Ext_F,[1 0 x]),matrix(Ext_F,[1 x x^2]),matrix(Ext_F,[1 0 0;0 1 0]),matrix(Ext_F,[1 0 0;0 1 x]),matrix(Ext_F,[1 0 x;0 1 x^2])];

In [49]:
overall_li = AbstractVector{AbstractVector{fpMatrix}}([])
for m in rep_matrices
    true_list = []
    QM = q_matroid_from_matrix(Ext_F,m)
    triples = Dim4_q_matroid_DFS(QM)
    for triple in triples
        answer = Are_q_matroid_dependentspaces(QM.field,triple[2])
        if answer == true
            push!(true_list,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[1],triple[2]])
        end
    end
    for tuple in true_list
        push!(overall_li,tuple[3])
    end
end     
overall_li    

594-element Vector{AbstractVector{fpMatrix}}:
 [[1 0 0 0; 0 0 0 1], [1 1 0 0; 0 0 0 1], [1 0 1 0; 0 0 0 1], [1 1 1 0; 0 0 0 1], [1 0 0 1; 0 1 0 1], [1 0 0 1; 0 0 1 1], [1 0 0 1; 0 1 1 1], [1 0 0 0; 0 1 0 1], [1 0 1 0; 0 1 0 1], [1 0 1 1; 0 1 0 1], [1 0 1 0; 0 1 1 1], [1 1 0 1; 0 0 1 1], [1 0 0 0; 0 0 1 1], [1 1 0 0; 0 0 1 1], [1 0 1 1; 0 1 1 1], [1 0 0 0; 0 1 1 1]]
 [[1 0 0 0; 0 0 0 1], [1 1 0 0; 0 0 0 1], [0 0 1 0; 0 0 0 1], [1 0 1 0; 0 0 0 1], [0 1 1 0; 0 0 0 1], [1 1 1 0; 0 0 0 1], [1 0 0 1; 0 1 0 1], [1 0 0 1; 0 0 1 0], [1 0 0 1; 0 0 1 1], [1 0 0 1; 0 1 1 0]  …  [0 1 0 1; 0 0 1 1], [1 0 1 1; 0 1 0 1], [1 1 0 1; 0 0 1 0], [1 0 1 0; 0 1 1 1], [1 0 1 1; 0 1 1 0], [1 1 0 1; 0 0 1 1], [1 0 0 0; 0 0 1 1], [1 1 0 0; 0 0 1 1], [1 0 1 1; 0 1 1 1], [1 0 0 0; 0 1 1 1]]
 [[1 0 0 0; 0 0 0 1], [0 1 0 0; 0 0 0 1], [1 1 0 0; 0 0 0 1], [0 0 1 0; 0 0 0 1], [1 0 1 0; 0 0 0 1], [0 1 1 0; 0 0 0 1], [1 1 1 0; 0 0 0 1], [1 0 0 1; 0 1 0 0], [1 0 0 1; 0 1 0 1], [1 0 0 1; 0 0 1 0]  …  [1 1 0 1; 0 0 1 0], [1

In [50]:
Isom_classes_from_basesV2(GF(2),overall_li)

10-element Vector{Any}:
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 16, 0)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 24, 6)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 28, 2)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 28, 0)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 34, 28)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 33, 183)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 32, 271)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 31, 87)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 30, 7)
 (Q-Matroid of rank 2 in 4-dim. vector-space over the Finite field of characteristic 2, 35, 0)

**The result is the following list of the form:**

***( ...| # bases-elm's | # elm's in isom-class )***

```
10-element Vector{Any}:
 (..., 16, 0)
 (..., 24, 6)
 (..., 28, 2)
 (..., 28, 0)
 (..., 34, 28)
 (..., 33, 183)
 (..., 32, 271)
 (..., 31, 87)
 (..., 30, 7)
 (..., 35, 0)
```

### Observation

**We see that the isomorphism-classes coming from the matrices leaking one class compared to those coming from the enumeration-algo., namely those q-matroids which have 31 bases-elm's.**

**These q-matroids are well known to be not representable.**

## Dimension 5
***

**We start by taking a q-matroid in dimension 4 and embed this one in dimension 5 by setting the fifth coordinate to 0.**

**So since we looking for 5-dim. q-matroids of rank 2 we have 14 possible options for this initial 4-dim. q-matroid to embed.**

**Again here we only looking at the possible isomorphism-classes and choose a representative.**

**Moreover since not all 4-dim. q-matroids are representable (there one not representable rank 2) we only state the number of bases elements of a representative q-matroid of an isomorphism-class:**

||||||||||||
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|rank 1|$15$|$14$|$12$|$8$|
|rank 2|$35$|$34$|$33$|$32$|$31$|$30$|$28_1$|$28_2$|$24$|$16$|

### Isomorphism-classes from matrices

**Before we do an example of the algorithm, we want to get an idea what kind of q-matroids we could expect.**

**Therefore we generate (2 $\times$ 5)-matrices with entries in $\mathbb{F}_{2^5}$, since we that every single of those matrices will give us a 5-dim. q-matroid.**

**(Of course only representable ones)**

**(Moreover in contrast to the 4-dim. case we not generate all possibilities, we rather choose random 10 elm's of "$2\times\mathbb{F}_{2^5}$")**

**After this we run an algorithm on the set of those matrices to identify all the different isomorphism-classes.**

**The results are given in the following list of the form:**

***( Rep.-Gen.-Mat. |...| # bases-elm's | # elm's in isom-class )***

```
19-element Vector{Any}:
 ([1 0 x^3 x^3 x^3; 0 1 x^3+1 x^4+x^3+x+1 x^4+x+1], ..., 146, 12659)
 ([1 0 x^3 x^3 x^3+1; 0 1 x^3+1 x^4+x^3+x+1 x^4+x^3+x+1], ..., 132, 2681)
 ([1 0 x^3 x^3 x^4+x^3+x+1; 0 1 x^3+1 x^4+x^3+x+1 x^4+x+1], ..., 150, 9317)
 ([1 0 x^3 x^3 x^4+x^3+x+1; 0 1 x^3+1 x^4+x^3+x+1 x^3+1], ..., 144, 2393)
 ([1 0 x^3 x^3 x^4+x+1; 0 1 x^3+1 x^4+x^3+x+1 x^3+x], ..., 152, 10361)
 ([1 0 x^3 x^3 x^3+x; 0 1 x^3+1 x^4+x^3+x+1 x^4+x^2+1], ..., 149, 4115)
 ([1 0 x^3 x^3 x^3+x; 0 1 x^3+1 x^4+x^3+x+1 0], ..., 151, 12389)
 ([1 0 x^3 x^3 x^4+x^3+x+1; 0 1 x^3+1 x^4+x+1 x^3+x], ..., 150, 3335)
 ([1 0 x^3 x^3 x^2+x; 0 1 x^3+1 x^2+x 0], ..., 151, 6311)
 ([1 0 x^3 x^3 x^3+1; 0 1 x^3+1 0 x^3+1], ..., 128, 365)
 ([1 0 x^3 x^3+1 x^4+x^3+x+1; 0 1 x^3+1 x^3+1 x^4+x+1], ..., 136, 4295)
 ([1 0 x^3 x^3+1 x^4+x^2+1; 0 1 x^3+1 x^3+1 x^4+x^2], ..., 112, 269)
 ([1 0 x^3 x^3+1 x^3; 0 1 x^3+1 x^3+1 x^3+1], ..., 96, 28)
 ([1 0 x^3 x^4+x^3+x+1 x^4+x^2+1; 0 1 x^3+1 x^3+x x^2+x], ..., 153, 1955)
 ([1 0 x^3 x^4+x^3+x+1 x^3+x; 0 1 x^3+1 x^4+x^2+1 x^4+x^2+1], ..., 150, 485)
 ([1 0 x^3 x^4+x+1 x^4+x^2+1; 0 1 x^3+1 x^2+x x^4+x^2], ..., 148, 1049)
 ([1 0 x^3 x^3+1 x^4+x^3+x+1; 0 1 x^4+x^3+x+1 x^4+x^3+x+1 x^4+x+1], ..., 140, 1691)
 ([1 0 x^3 x^3+1 x^3; 0 1 x^4+x^3+x+1 x^4+x^3+x+1 x^4+x^3+x+1], ..., 112, 132)
 ([1 0 x^3 x^4+x^3+x+1 x^3+x; 0 1 0 0 0], ..., 120, 239)
```

**We give two other examples, using a smaller field extension:**

**Extension $m=2$:**

```
Ext_F,x = FiniteField(2,2,"x")
mats = matrix_collec(Ext_F,2,5);
isoms = Isom_classes_from_matsV2(Ext_F,mats)
```

**Result:**
```
3-element Vector{Any}:
 ([1 0 0 0 0; 0 1 1 x x+1], Q-Matroid of rank 2 in 5-dim. vector-space over the Finite field of characteristic 2, 96, 1343)
 ([1 0 0 0 x; 0 1 1 x x+1], Q-Matroid of rank 2 in 5-dim. vector-space over the Finite field of characteristic 2, 120, 2687)
 ([1 0 0 0 1; 0 1 1 0 0], Q-Matroid of rank 2 in 5-dim. vector-space over the Finite field of characteristic 2, 64, 63)
```

**Extension $m=3$:**

```
Ext_F,x = FiniteField(2,3,"x")
mats = matrix_collec(Ext_F,2,5)
isoms = Isom_classes_from_matsV2(Ext_F,mats)
```

**Result:**

### Enumeration-Algorithm

**Here we skip the list of all possible inputs and their respective outputs.**

**We only collect all these different q-matroids in one list and run an algorithm to distinguish the different isomorphism-classes.**

**For computing we use the following algorithm on the server:**
```
bases31 = AbstractVector{fpMatrix}([])
two_spaces = subspaces_fix_dim(GF(2),2,4)
A = matrix(GF(2),[1 0 0 0;0 1 0 0])
B = matrix(GF(2),[0 0 1 0;0 0 0 1])
C = matrix(GF(2),[1 0 0 1;0 1 1 0])
D = matrix(GF(2),[1 0 1 0;0 1 1 1])
l = [A,B,C,D]
for space in two_spaces
    if !(space in l)
        push!(bases31,space)
    end
end
qm31 = Q_Matroid(GF(2),bases31)
```
```
Ext_F,x = FiniteField(2,4,"x")
qmats = [matrix(Ext_F,[1 0 0 0]),
                matrix(Ext_F,[1 x 0 0]),
                matrix(Ext_F,[1 x x^2 0]),
                matrix(Ext_F,[1 x x^2 x^3]),
                matrix(Ext_F,[1 0 0 0; 0 1 1 0]),
                matrix(Ext_F,[1 0 0 0; 0 1 1 x]),
                matrix(Ext_F,[1 0 0 x; 0 1 1 x^2]),
                matrix(Ext_F,[1 0 0 0; 0 1 x x^2]),
                matrix(Ext_F,[1 0 0 x^2+x; 0 1 x^2+x x^2+x+1]),
                matrix(Ext_F,[1 0 0 x; 0 1 x x^2]),
                matrix(Ext_F,[1 0 0 x^2; 0 1 x x^3]),
                matrix(Ext_F,[1 0 0 x; 0 1 x x^3]),
                matrix(Ext_F,[1 0 x x^2; 0 1 x^2 x^2+x]),
                qm31
                ]
```
```
overall_li = AbstractVector{AbstractVector{fpMatrix}}([])
for i in range(1,length(qmats))
    if i != length(qmats)
        true_list = []
        QM = q_matroid_from_matrix(Ext_F,qmats[i])
        triples = Dim5_q_matroid_DFS(QM)
        for triple in triples
            answer = Are_q_matroid_dependentspaces(QM.field,triple[2])
            if answer == true
                push!(true_list,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[1],triple[2]])
            end
        end
        for tuple in true_list
            push!(overall_li,tuple[3])
        end
    elseif i == length(qmats)
        QM = qmats[i]
        triples = Dim5_q_matroid_DFS(QM)
        for triple in triples
            answer = Are_q_matroid_dependentspaces(QM.field,triple[2])
            if answer == true
                push!(true_list,[length(triple[3]),length(triple[1])+length(triple[2]),triple[3],triple[1],triple[2]])
            end
        end
        for tuple in true_list
            push!(overall_li,tuple[3])
        end
    end
end     
```
```
Isom_classes_from_basesV2(GF(2),overall_li)
```